resource-managers/kubernetes/core/src/main/scala/org/apache/spark/shuffle/KubernetesLocalDiskShuffleExecutorComponents.scala

```diff
@@ -84,24 +87,46 @@ object KubernetesLocalDiskShuffleExecutorComponents extends Logging {
           .flatMap(_.listFiles).filter(_.isDirectory) // executor-xxx
           .flatMap(_.listFiles).filter(_.isDirectory) // blockmgr-xxx
           .flatMap(_.listFiles).filter(_.isDirectory) // 00
-          .flatMap(_.listFiles).filterNot(_.getName.contains(".checksum"))
+          .flatMap(_.listFiles)
         if (files != null) files.toSeq else Seq.empty
       }
+      .partition(_.getName.contains(".checksum"))
+    val (indexFiles, dataFiles) = files.partition(_.getName.endsWith(".index"))
 
-    logInfo(s"Found ${files.size} files")
+    logInfo(s"Found ${dataFiles.size} data files, ${indexFiles.size} index files, " +
+        s"and ${checksumFiles.size} checksum files.")
+
+    // Build hashmaps for faster access with data file name as a key
+    val checksumFileMap = new mutable.HashMap[String, File]()
+    val algorithm = conf.get(SHUFFLE_CHECKSUM_ALGORITHM)
+    checksumFiles.foreach { f =>
+      checksumFileMap.put(f.getName.replace(".checksum." + algorithm, ".data"), f)
+    }
+    val indexFileMap = new mutable.HashMap[String, File]()
+    indexFiles.foreach { f =>
+      indexFileMap.put(f.getName.replace(".index", ".data"), f)
+    }
 
     // This is not used.
     val classTag = implicitly[ClassTag[Object]]
     val level = StorageLevel.DISK_ONLY
-    val (indexFiles, dataFiles) = files.partition(_.getName.endsWith(".index"))
+    val checksumDisabled = !conf.get(SHUFFLE_CHECKSUM_ENABLED)
     (dataFiles ++ indexFiles).foreach { f =>
       logInfo(s"Try to recover ${f.getAbsolutePath}")
       try {
         val id = BlockId(f.getName)
         // To make it sure to handle only shuffle blocks
         if (id.isShuffle) {
-          val decryptedSize = f.length()
-          bm.TempFileBasedBlockStoreUpdater(id, level, classTag, f, decryptedSize).save()
+          // For index files, skipVerification is true and checksumFile and indexFile are ignored.
+          val skipVerification = checksumDisabled || f.getName.endsWith(".index")
+          val checksumFile = checksumFileMap.getOrElse(f.getName, null)
+          val indexFile = indexFileMap.getOrElse(f.getName, null)
+          if (skipVerification || verifyChecksum(algorithm, id, checksumFile, indexFile, f)) {
+            val decryptedSize = f.length()
+            bm.TempFileBasedBlockStoreUpdater(id, level, classTag, f, decryptedSize).save()
+          } else {
+            logInfo(s"Ignore ${f.getAbsolutePath} due to the verification failure.")
+          }
```

Comment:
Hmm, if checksum validation fails, it just ignores the file instead of stoping recovery?

Reply:
It's because the checksum is a validation per file. We can still recover the healthy files.

Reply:
The corrupted files might be a result of some un-flushed files. So, we still need to try to recover the other successfully written files. It's consistent with the existing behavior.

Reply:
We use `Cause.DISK_ISSUE` here before.
https://github.com/apache/spark/blob/9520087409d5bd7e6a2651dacf2c295d564d5559/common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/checksum/ShuffleChecksumHelper.java#L140-L146

Reply:
> It's consistent with the existing Spark behavior.

Hmm, where the behavior is? Could you point me to the place?

In `ShuffleBlockFetcherIterator`, Spark fetches corrupted block twice and uses checksum to diagnose the cause. Otherwise, I'm not sure if another place uses checksum.

Reply:
Yes, what I referred was that diagnosis part where we proceed to `FetchFailedException` in order to make resubmit the previous stage. We don't immediately exclude the node itself.

Reply:
The stage will be re-tried 5 times (attemptId: 0 ~ 4) and the node exclusion is handled by independent configurations like `spark.excludeOnFailure.enabled` (and its subconfigs) or `spark.stage.ignoreDecommissionFetchFailure` (to ignore in some context).

Reply:
But here this is recovery of shuffle data from PVC, it won't proceed to any resubmission of previous stage. I'm not sure is, if we simply ignore any checksum failure. What's the purpose to validate checksum here?

Reply:
Ur, it seems that you assume that this PR should be a on-off switch for the whole physical volume health-check. Ya, it could be. However, it's not the goal for now. The main request I received was to detect the potential OS corruptions where some unflushed files are visible accidentally during detaching and re-attaching.

Apache Spark shuffle data are only meaningful inside the relationship of `index file`, `data file`, and `checksum file`. And, `checksum` is designed as an auxiliary file for additional analysis instead of mandatory files.

So, the goal of this PR is simple in the context of data recovery.
- If there is a checksum, we try to verify the data once more to make it sure there is no corruption because this is a remounted data. If it passes, we are good to use the verified data.
- If there is a suspicious part, we discard those data while keeping recovering the other shuffle data. If we discard, it will be recomputed by Apache Spark like the case we didn't remount this PVC. So, there is no issue by discard those  data when checksum failures happen.

Just to give a more background, although we assume EBS or EFS in the community, there are many K8s PV implementations and some of them are new and proprietary implementations which we don't know what is going on. This is proposed as a moderated protected mode.

Reply:
> if we simply ignore any checksum failure. What's the purpose to validate checksum here?

When we ignore, it's not recovered because we don't call `TempFileBasedBlockStoreUpdater`. The ignored files are in the old (and some random local directory) path which the current Spark BlockManager is not aware of.
```scala
val decryptedSize = f.length()
bm.TempFileBasedBlockStoreUpdater(id, level, classTag, f, decryptedSize).save()
```

Reply:
> If there is a suspicious part, we discard those data while keeping recovering the other shuffle data files. If we discard, it will be recomputed by Apache Spark automatically like the case we didn't remount this PVC at all. So, there is no issue by discarding those suspicious data when checksum failures happen.

Oh okay. If, for example, one shuffle file is not recovered, will there be a fetch failure exception like normal shuffle fetch case and that triggers to recompute? Is this what you mean?

Reply:
Yes~ It's a normal behavior. We don't assume that the previous stage itself is completed or not. In general, the executor loss can happen at any time during the processing. So, recompute will happen in most cases. PVC is only trying to recover what Spark got before (even the data is a partial computation result).

Reply:
Okay, makes sense.